[AArch64] Initial sched model for Neoverse V3, V3AE

simonwallis2 opened the pull request:
Add the scheduling models for Neoverse V3 and Neoverse V3AE 
based on information taken from the V3 Software Optimization guide: 
https://developer.arm.com/documentation/109678/300/?lang=en

and on information taken from the V3AE Software Optimization guide: 
https://developer.arm.com/documentation/109703/300/?lang=en

Implements https://github.com/llvm/llvm-project/issues/134977

Commit:
[AArch64] Initial sched model for Neoverse V3, V3AE

Add the scheduling models for Neoverse V3 and Neoverse V3AE
based on information taken from the V3 Software Optimization guide:
https://developer.arm.com/documentation/109678/300/?lang=en

and on information taken from the V3AE Software Optimization guide:
https://developer.arm.com/documentation/109703/300/?lang=en

Implements https://github.com/llvm/llvm-project/issues/134977

Change-Id: I7fede22330c1bce14e9bb978543b3f8b7f43162d

davemgreen commented on llvm/lib/Target/AArch64/AArch64SchedNeoverseV3.td:
```diff
@@ -0,0 +1,2781 @@
+//===----------------------------------------------------------------------===//
+
+def NeoverseV3Model : SchedMachineModel {
+  let IssueWidth            =   8; // Expect best value to be slightly higher than V2
```
Should this be 10?

c-rhodes reviewed (commented):
thanks Simon, not finished reviewing yet but left some initial comments

Asher8118 reviewed (commented):
Thanks for the patch! I have added a few comments. I have only looked through the V3-basic-instruction.s so far. I'll have to look through the NEON and SVE as well as the remaining tests.

simonwallis2 commented on llvm/test/tools/llvm-mca/AArch64/Neoverse/V3-basic-instructions.s:
```diff
@@ -0,0 +1,3783 @@
+# CHECK-NEXT:  1      1     0.13                        sub	sp, sp, #288
+# CHECK-NEXT:  1      1     0.13                        sub	wsp, w19, #16
+# CHECK-NEXT:  1      1     0.13                        adds	w13, w23, #291, lsl #12
+# CHECK-NEXT:  1      1     0.13                        cmn	w2, #4095
```
Thanks. Agreed. This was fixed by changing the count of UnitFlgs from 8 to 4.

simonwallis2 commented on llvm/test/tools/llvm-mca/AArch64/Neoverse/V3-basic-instructions.s:
```diff
@@ -0,0 +1,3783 @@
+# CHECK-NEXT:  1      1     0.13                        adc	xzr, x3, x4
+# CHECK-NEXT:  1      1     0.13                        adc	x9, xzr, x10
+# CHECK-NEXT:  1      1     0.13                        adc	x20, x0, xzr
+# CHECK-NEXT:  1      1     0.13                        adcs	w29, w27, w25
```
Thanks. Agreed. This was fixed by changing the count of UnitFlgs from 8 to 4.

simonwallis2 commented on llvm/test/tools/llvm-mca/AArch64/Neoverse/V3-basic-instructions.s:
```diff
@@ -0,0 +1,3783 @@
+# CHECK-NEXT:  1      1     0.13                        cmn	w3, w5
+# CHECK-NEXT:  1      1     0.13                        adds	w20, wzr, w4
+# CHECK-NEXT:  1      1     0.13                        adds	w4, w6, wzr
+# CHECK-NEXT:  1      1     0.13                        adds	w11, w13, w15
```
I have not yet addressed this.  It was not affected by changing the count of Flags from 8 to 4.

Commit:
[AArch64] Initial sched model for Neoverse V3, V3AE

Add the scheduling models for Neoverse V3 and Neoverse V3AE
based on information taken from the V3 Software Optimization guide:
https://developer.arm.com/documentation/109678/300/?lang=en

and on information taken from the V3AE Software Optimization guide:
https://developer.arm.com/documentation/109703/300/?lang=en

Implements https://github.com/llvm/llvm-project/issues/134977

Change-Id: Ib43fb98dda9e8a7cdf8c8910df63c89936701ff5

c-rhodes commented on llvm/test/tools/llvm-mca/AArch64/Neoverse/V3-basic-instructions.s:
```diff
@@ -2532,18 +2532,18 @@ drps
-# CHECK-NEXT:  1      0     0.13                        mov	x3, xzr
-# CHECK-NEXT:  1      0     0.13                        mov	wzr, w2
-# CHECK-NEXT:  1      0     0.13                        mov	w3, w5
+# CHECK-NEXT:  1      0     0.10                        mov	x3, x6
```
this is now too high?

simonwallis2 replied:
I think it's OK.
4.11 Zero Latency MOVs A subset of register-to-register move operations and move immediate operations are executed with zero latency. These instructions do not utilize the scheduling and execution resources of the machine. 
So in the V3 model, mov x3,x6 is constrained only by the IssueWidth, currently 10.

simonwallis2 commented on llvm/lib/Target/AArch64/AArch64Processors.td:
```diff
@@ -1295,9 +1295,9 @@ def : ProcessorModel<"neoverse-v1", NeoverseV1Model,
 def : ProcessorModel<"neoverse-v2", NeoverseV2Model,
                      ProcessorFeatures.NeoverseV2, [TuneNeoverseV2]>;
-def : ProcessorModel<"neoverse-v3", NeoverseV2Model,
+def : ProcessorModel<"neoverse-v3", NeoverseV3Model,
```
There is now a new premerge checks failire.
Next revision of the patch will include updated llvm-mca/AArch64/Cortex/X4-sve-instructions.s

Commit:
[AArch64] Initial sched model for Neoverse V3, V3AE

Add the scheduling models for Neoverse V3 and Neoverse V3AE
based on information taken from the V3 Software Optimization guide:
https://developer.arm.com/documentation/109678/300/?lang=en

and on information taken from the V3AE Software Optimization guide:
https://developer.arm.com/documentation/109703/300/?lang=en

Implements https://github.com/llvm/llvm-project/issues/134977

Change-Id: I8c938c2a49b16531e82e919cc68db77e015febf2

Commit:
[AArch64] Initial sched model for Neoverse V3, V3AE

Add the scheduling models for Neoverse V3 and Neoverse V3AE
based on information taken from the V3 Software Optimization guide:
https://developer.arm.com/documentation/109678/300/?lang=en

and on information taken from the V3AE Software Optimization guide:
https://developer.arm.com/documentation/109703/300/?lang=en

Implements https://github.com/llvm/llvm-project/issues/134977

Change-Id: Ica2d4c5d869553508feeee17acb4feec68f39aba

c-rhodes reviewed (approved):
LGTM cheers

it is a bit of a shame there's so much duplication when these models are more common than not, most of the differences seems to stem from V3AE having 2 less vector pipes, but I guess that also applies across other neoverse models and fixing it would take a fair bit of refactoring

Asher8118 reviewed (commented):
Thanks for updating the patch! A few more comments on the SVE instructions:

- The SWOG shows the throughput of SVE PTRUES to be 2, but it [currently shows as 1](https://github.com/llvm/llvm-project/blob/73a42dda79f0d4e4ac4caa60b7738aeeb508994e/llvm/test/tools/llvm-mca/AArch64/Neoverse/V3-sve-instructions.s#L5361).
- The V3 SWOG shows that SVE Dot product, 16 bit has execution latency of 3, which [currently shows as 4](https://github.com/llvm/llvm-project/blob/73a42dda79f0d4e4ac4caa60b7738aeeb508994e/llvm/test/tools/llvm-mca/AArch64/Neoverse/V3-sve-instructions.s#L5475).
- Multiply accumulate B,H,S element size shows throughput of 2 in the SWOG, but [currently shows as 1](https://github.com/llvm/llvm-project/blob/73a42dda79f0d4e4ac4caa60b7738aeeb508994e/llvm/test/tools/llvm-mca/AArch64/Neoverse/V3-sve-instructions.s#L4953).
- STNT1B, vector+scalar 32-bit element has execution throughput of 1/3 in the SWOG, [but shows as 1/2 currently](https://github.com/llvm/llvm-project/blob/73a42dda79f0d4e4ac4caa60b7738aeeb508994e/llvm/test/tools/llvm-mca/AArch64/Neoverse/V3-sve-instructions.s#L6185C23-L6185C34). That's the instruction that caught my eye, but there might be more store instructions showing the wrong throughput. I saw there's differences here between the V2 and V3 SWOG (eg: scatter stores) which is what made me check this instruction. I'd double check those match the V3 SWOG.

Asher8118 reviewed (commented):
I've looked over the ASIMD instructions and clear-upper-regs test. Both look good to me. Other than my previous comments regarding SVE instructions, I didn't notice any other discrepancies with the SWOG.

Commit:
[AArch64] Initial sched model for Neoverse V3, V3AE

Add the scheduling models for Neoverse V3 and Neoverse V3AE
based on information taken from the V3 Software Optimization guide:
https://developer.arm.com/documentation/109678/300/?lang=en

and on information taken from the V3AE Software Optimization guide:
https://developer.arm.com/documentation/109703/300/?lang=en

Implements https://github.com/llvm/llvm-project/issues/134977

Change-Id: I2355c8a92c2350d55b670d6a2acb0e22e1cacc54

simonwallis2 commented:
> Thanks for updating the patch! A few more comments on the SVE instructions:
> 
> * The SWOG shows the throughput of SVE PTRUES to be 2, but it [currently shows as 1](https://github.com/llvm/llvm-project/blob/73a42dda79f0d4e4ac4caa60b7738aeeb508994e/llvm/test/tools/llvm-mca/AArch64/Neoverse/V3-sve-instructions.s#L5361).
ptrues has been fixed for both V3 and V3AE and removed unused patterns.

> * The V3 SWOG shows that SVE Dot product, 16 bit has execution latency of 3, which [currently shows as 4](https://github.com/llvm/llvm-project/blob/73a42dda79f0d4e4ac4caa60b7738aeeb508994e/llvm/test/tools/llvm-mca/AArch64/Neoverse/V3-sve-instructions.s#L5475).
sve dot product has been fixed for both V3 and V3AE. Also complex dot product 16-bit element.

> * Multiply accumulate B,H,S element size shows throughput of 2 in the SWOG, but [currently shows as 1](https://github.com/llvm/llvm-project/blob/73a42dda79f0d4e4ac4caa60b7738aeeb508994e/llvm/test/tools/llvm-mca/AArch64/Neoverse/V3-sve-instructions.s#L4953).
multiply accumulate has been fixed for both V3 and V3AE.

> * STNT1B, vector+scalar 32-bit element has execution throughput of 1/3 in the SWOG, [but shows as 1/2 currently](https://github.com/llvm/llvm-project/blob/73a42dda79f0d4e4ac4caa60b7738aeeb508994e/llvm/test/tools/llvm-mca/AArch64/Neoverse/V3-sve-instructions.s#L6185C23-L6185C34). That's the instruction that caught my eye, but there might be more store instructions showing the wrong throughput. I saw there's differences here between the V2 and V3 SWOG (eg: scatter stores) which is what made me check this instruction. I'd double check those match the V3 SWOG.
Did not need changing for V3AE. 
For V3, the last 10 rows in the SVE Store Instructions table have been fixed, i.e. all of the scatter stores, 32-bit element size and 64-bit element size.

Thanks for the thorough and detailed review.

Asher8118 reviewed (approved):
Thanks for the updates ! LGTM 👍🏻
